File: LeetCode/121. Best Time to Buy and Sell Stock.py
# 单次遍历即可解决，保存到当前为止的借个最小值以及收益的最大值，两个值不断更新即可

class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        if (len(prices) == 0):
            return 0
        max_profit = 0
        min_price = None
        
        for price in prices:
            if (min_price is not None):
                profit = price - min_price
                if (profit > max_profit):
                    max_profit = profit
            if (min_price is None or price < min_price):
                min_price = price
        return max_profit
            

File: LeetCode/121. Best Time to Buy and Sell Stock.cpp
// 记录前面最小一个数，然后遍历时更新这个最小值，并且判断是否能获得更大利润即可
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        if (prices.size() < 2) {
            return 0;
        }
        int minStart = prices[0];
        int res = 0;
        for (int i = 1; i < prices.size(); ++i) {
            res = max(res, prices[i] - minStart);
            minStart = min(minStart, prices[i]);
        }
        return res;
    }
};